compiler/src/iree/compiler/Codegen/LLVMGPU/AMDGPUChainedMatmulPass.cpp

```diff
@@ -13,70 +13,121 @@ namespace mlir::iree_compiler {
 
 namespace {
 
+/// Let's assume that we only have vector.contract with the standard indexing
+/// maps:
+///    (m, n, k), A: (m, k), B: (k, n), C: (m, n).
+/// We will represent this contract operation by a "@".
+///
+/// Given a matmul:
+///
+/// C = A @ B
+///
+/// This pass decides when to convert this matmul to:
+///
+/// A.T = transpose(A)
+/// B.T = transpose(B)
+/// C.T = A.T @ B.T
+/// C = transpose(C.T)
+///
+/// This is useful when the "@" instruction that the hardware lowers to
+/// has a specific layout (see VectorLayoutInterface for more information)
+/// but the further uses of C expects a transposed layout to the produced
+/// layout.
+///
+/// For example, for "@" lowering to AMDGPU MFMA instructions, the operands
+/// have layout L and L.T and the result has the layout L.T .
+/// So if you have a chain of matmuls:
+///
+/// C (L.T) = A (L) @ B (L.T)
+/// E (L.T) = C (L.T)  @ D (L.T)
+///            ^^^^^^^
+///            Expected layout by instruction is L
+///
+/// To fix this, we can apply this transformation on the first matrix:
+///
+/// C.T (L.T) = A.T (L) @ B (L.T)
```

Comment:
Same for this
```
C.T (L.T) = B.T (L) @ A.T (L.T)
```

Reply:
i looked at these transposes too much and my brain got friend. Thanks for checking :)